sonic_installer/main.py

```diff
@@ -1,54 +1,52 @@
 #! /usr/bin/python -u
 
 import os
-import re
-import signal
 import sys
 import time
 import click
 import urllib
 import syslog
 import subprocess
 from swsssdk import SonicV2Connector
-import collections
-import platform
-
-HOST_PATH = '/host'
-IMAGE_PREFIX = 'SONiC-OS-'
-IMAGE_DIR_PREFIX = 'image-'
-ONIE_DEFAULT_IMAGE_PATH = '/tmp/sonic_image'
-ABOOT_DEFAULT_IMAGE_PATH = '/tmp/sonic_image.swi'
-IMAGE_TYPE_ABOOT = 'aboot'
-IMAGE_TYPE_ONIE = 'onie'
-ABOOT_BOOT_CONFIG = '/boot-config'
-BOOTLOADER_TYPE_GRUB = 'grub'
-BOOTLOADER_TYPE_UBOOT = 'uboot'
-ARCH = platform.machine()
-BOOTLOADER = BOOTLOADER_TYPE_UBOOT if ("arm" in ARCH) or ("aarch64" in ARCH) else BOOTLOADER_TYPE_GRUB
+
+from .common import run_command
+
+from .bootloader.aboot import AbootBootloader
+from .bootloader.grub import GrubBootloader
+from .bootloader.uboot import UbootBootloader
```

Comment:
You can import many in `bootloader.__init__.py`, and here `from bootloader import AbootBootloader, GrubBootloader, UbootBootloader`.

The benefit is hiding the file structure inside subdir.

Reply:
Joe seemed like he was not too keen on having code in `__init__.py`.
If he is fine with such a change I will do it and likely move the `BOOTLOADERS` definition there too.

Reply:
I am opposed to defining classes in the \_\_init\_\_.py file. As I stated above, the primary purpose of the \_\_init\_\_.py file is to initialize the package, not to hold content. Qi's request here is reasonable. This person's description aligns with my thoughts on the usage of \_\_init\_\_.py: https://mikegrouchy.com/blog/be-pythonic-__init__py

Reply:
The `get_bootloader` function looks good to me.

There is a good reference if related: [https://mehulkar.com/blog/2017/04/the-factory-pattern-and-circular-dependencies/](https://mehulkar.com/blog/2017/04/the-factory-pattern-and-circular-dependencies/)

---
In reply to: [443153452](https://github.com/Azure/sonic-utilities/pull/953#discussion_r443153452) [](ancestors = 443153452)

Reply:
Sure I'll move some of the declaration to __init__.py

Regarding your link Qi, I personally disagree with the logic used by this post.
For me a Base class should never have a function that refers and import all its base classes.
IMO it's bad design, I much prefer self registering class over this approach.
Also this example is simplisistic now you have 1M animals, good luck with your 2M line function.
If the logic to detect each animal gets complicated now you have a huge spaghetti mess function that offloads the animal specific detection to the base class which I would never agree to in a review.

I'll do another pass and you guys can comment on my new approach.